cocos/core/gfx/base/define.ts

```diff
@@ -953,19 +954,44 @@ export class Color {
     }
 }
 
+/**
+ * For non-vulkan backends, to maintain compatibility and maximize
+ * descriptor cache-locality, descriptor-set-based binding numbers need
+ * to be mapped to backend-specific bindings based on maximum limit
+ * of available descriptor slots in each set.
+ *
+ * The GFX layer assumes the binding numbers for each descriptor type inside each set
+ * are guaranteed to be consecutive, so the mapping procedure is reduced
+ * to a simple shifting operation. This data structure specifies the
+ * capacity for each descriptor type in each set.
+ *
+ * The `setIndices` field defines the binding ordering between different sets.
+ * The last set index is treated as the 'flexible set', whose capacity is dynamically
+ * assigned based on the total available descriptor slots on the runtime device.
+ */
 export class BindingMappingInfo {
     declare private _token: never; // to make sure all usages must be an instance of this exact class, not assembled from plain object
 
     constructor (
-        public bufferOffsets: number[] = [],
-        public samplerOffsets: number[] = [],
-        public flexibleSet: number = 0,
+        public maxBlockCounts: number[] = [0],
+        public maxSamplerTextureCounts: number[] = [0],
+        public maxSamplerCounts: number[] = [0],
+        public maxTextureCounts: number[] = [0],
+        public maxBufferCounts: number[] = [0],
+        public maxImageCounts: number[] = [0],
+        public maxSubpassInputCounts: number[] = [0],
+        public setIndices: number[] = [0],
```

Comment:
Looks like a lot more data, have memory impact ?

Reply:
And I don't understand why maxXXXCounts are arrays...

Reply:
The limits are set per descriptor set, thus the array structure.
The arrays are very small by nature, 4 elements at most according to vk spec. For our purposes this is fixed as length-3 arrays, corresponding to the 3 descriptor sets (global, material & local). 

This interface is transparent to the upper layer most of the time, and the rule is pretty straightforward. 
The complexities are all encapsulated inside the actual backends.